crates/libcontainer/src/utils.rs

```diff
@@ -379,4 +401,38 @@ mod tests {
 
         Ok(())
     }
+
+    // the following test is marked as serial because
+    // we are doing unshare of user ns and fork, so better to run in serial,
+    #[test]
+    #[serial]
+    fn test_userns_spec_validation() -> Result<(), test_utils::TestError> {
+        use nix::sched::{unshare, CloneFlags};
+        // default rootful spec
+        let rootful_spec = Spec::default();
+        // as we are not in a user ns, and spec does not have user ns
+        // we should get error here
+        assert!(validate_spec_for_new_user_ns(&rootful_spec).is_err());
+
+        let rootless_spec = Spec::rootless(1000, 1000);
+        // because the spec contains user ns info, we should not get error
+        assert!(validate_spec_for_new_user_ns(&rootless_spec).is_ok());
+
+        test_utils::test_in_child_process(|| {
+            unshare(CloneFlags::CLONE_NEWUSER).unwrap();
+            // here we are in a new user namespace
+            let rootful_spec = Spec::default();
+            // because we are already in a new user ns, it is fine if spec
+            // does not have user ns, and because the test is running as
+            // non root
+            assert!(validate_spec_for_new_user_ns(&rootful_spec).is_ok());
+
+            let rootless_spec = Spec::rootless(1000, 1000);
+            // following should succeed irrespective if we're in user ns or not
+            assert!(validate_spec_for_new_user_ns(&rootless_spec).is_ok());
+            Ok(())
+        })?;
+
+        Ok(())
+    }
```

Comment:
This still doesn't cover all possible permutations, but as we should run the tests themselves as non-root, we cannot test cases where the user is actually root, and both specs pass.